runtime/src/comm/gasnet/comm-gasnet.c

```diff
@@ -836,13 +836,38 @@ static void set_num_comm_domains() {
 #endif
 }
 
-void chpl_comm_pre_topo_init(void) {
-  // not supported on this platform
-  chpl_set_num_locales_on_node(1);
+void chpl_comm_pre_topo_init(int *argc_p, char ***argv_p) {
+  // Initialize gasnet so that we can call gex_System_QueryHostInfo and
+  // set the number of locales on our node which allows the rest of the
+  // runtime to determine which cores this locale will use. gasnet_attach
+  // is called in chpl_comm_init which is called after the cores have been
+  // determined.
+  gex_Rank_t      infoCount;
+  gex_Rank_t      myIndex;
+
+  set_max_segsize();
+  set_num_comm_domains();
```

Comment:
This calls `chpl_topo_getNumCPUsPhysical()`. Can that be called pre topo init?

Reply:
I'm not sure how to break this circularity when `CHPL_COMM_SUBSTRATE=aries`.  We want to create one domain per core and to do so we need to set `GASNET_DOMAIN_COUNT` before calling `gasnet_init`, but we won't know how many cores are available to the locale until after `gasnet_init` is called. I suppose I could call `gasnet_init` with `GASNET_DOMAIN_COUNT=1`, get the number of cores, then call `gasnet_init` again with `GASNET_DOMAIN_COUNT` set to the proper value. Or I could set `GASNET_DOMAIN_COUNT` to a maximum and we don't use some domains. I don't know implications of not using all the domains.

Reply:
We can't call `gasnet_init` twice. Setting it to a max could use more resources than ideal. Can we set it based on max accessible cores, ignoring any reserved by the runtime? That should be a good enough value.

Reply:
Aries only supports 120 or so communication domains (I forget the exact number) so we probably can't just give each process numCoresPerNode comm domains.

In general, it's not super important, especially as XC ages. It's nice to have as a backup for users if they hit bugs with ugni and nice for us to do performance and correctness comparisons against, but it's not critical. Longer term though I expect more gasnet configurations to need to know the number of cores (for explicit multi-endpoint support, where the `GASNET_DOMAIN_COUNT` enables implicit endpoints that gasnet round-robins too under the covers)

Reply:
I changed `set_num_comm_domains` for aries to determine the maximum number of cores the locale could have based on the total number of cores and the number of locales on the node. This may overestimate because of inaccessible cores, which we don't know about at this point in the initialization. Let me know what you think.